include/envoy/network/address.h

```diff
@@ -161,6 +161,25 @@ class Instance {
    * @return the type of address.
    */
   virtual Type type() const PURE;
+
+  /**
+   * @return a pointer to the POSIX sockaddr struct associated with the address instance.
+   * Depending on the type() of address, this maps to either a sockaddr_in/sockaddr_in6/sockaddr_un
+   *
+   * This API is used when the caller has the address instance available and requires to
+   * provide the sockaddr parameter to a posix API.
+   */
+  virtual const sockaddr* sockAddr() const PURE;
```

Comment:
In general we shouldn't be leaking these type of UNIX details out of the address API IMO. @danzh2010 what is your thought here in terms of how we want to abstract this, especially WRT to the IO handle refactor? From looking further maybe this just doesn't need to be in the public interface?

Reply:
Sure, I can move it inside iosocketimpl. But the question remains on how to expose the sockaddr from the AddressInstance. 

I am thinking of moving the 2 methods from Instance interface to BaseInstance (inside addressimpl.h) and expose a method in AddressImpl.h to return the sockaddr from a const instance like this:
`const sockaddr* sockAddrFromInstance(InstanceConstSharedPtr address);`

Caller can do this:
`  const sockaddr* server_addr = Address::sockAddrFromInstance(server_socket->localAddress());`

Reply:
We don't want to expose sockaddr. The vision is wherever we need sockaddr to make a system call now will be replaced with IoHandle::someSysCallWrapper(Address& address). In the IoSocketHandleImpl::someSysCallWrapper() we can use sockaddr.

Reply:
Right.. my question is how we can construct a sockaddr object from an address inside IoSocketHandleImpl::someSysCallWrapper()? Since the sockaddr lives inside the address, I am trying to see if it can be harnessed somehow - rather than writing the logic to construct the sockaddr all over again.